### DISCUSSION TITLE
Error after updating Moose

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#22920] [Q&A Getting Started] [answered]
Hello:

Due to a "missing symbol" problem with my app, I decided to eliminate some possibilities and so updated my Moose environment and Moose itself using the instructions from Getting Started.  Namely:
```
conda activate moose
conda update --all

cd ~/projects/moose
git fetch origin
git rebase origin/master

Then, I tried to run the tests.  I got more than 50 failures, and so ran

git clean -xfd

Upon trying to run the tests, I then got this error:

(moose) moose % cd test 
(moose) test % ./run_tests -j12
Building and linking hit...
make[1]: Entering directory '/Users/joe/projects/moose/framework/contrib/hit'
Building hit for python with python3-config
mpicxx -std=c++17 -g -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/joe/mambaforge3/envs/moose/include -std=c++17 main.cc parse.cc lex.cc braceexpr.cc -o hit
make[1]: Leaving directory '/Users/joe/projects/moose/framework/contrib/hit'
Error! Could not find libmesh_config.h in any of the usual locations!
```
This is on an intel Mac, running 12.6.1.

Help please!  Thanks,
   Joe Kelly

**Comment by GiudGiud** [comment]
Hi Joe

I'd be worried that git clean -xfd removed the moose install. Did you remake moose after updating libmesh?
The missing libmesh_config.h is pointing to a missing libmesh but that could be a red herring. 

I would also check the version of libmesh with
` mamba list | grep moose`
then I would rebuild moose in moose/test

Guillaume

**Comment by joe61vette** [comment]
Hi Guillaume:

After running 'git clean -xfd' (as suggested in another discussion), I then went to /test and entered ./run_tests.  So I did not explicitly enter the make command.

Also:

```bash
(moose) test % mamba list | grep moose
# packages in environment at /Users/joe/mambaforge3/envs/moose:
moose-libmesh             2022.10.05              build_0    https://conda.software.inl.gov/public
moose-libmesh-vtk         9.1.0               he71d177_12    https://conda.software.inl.gov/public
moose-mpich               4.0.2                   build_2    https://conda.software.inl.gov/public
moose-petsc               3.16.6                  build_1    https://conda.software.inl.gov/public
moose-tools               2022.07.18      py310hcffbc79_0    https://conda.software.inl.gov/public
```

Thanks for the help,
  Joe

**Reply by GiudGiud** [reply]
The moose-libmesh and moose-tools are a little on the older side. I have:

```
moose-libmesh             2022.11.07              build_0    https://conda.software.inl.gov/public
moose-libmesh-vtk         9.1.0               h829a923_14    https://conda.software.inl.gov/public
moose-mpich               4.0.2                   build_4    https://conda.software.inl.gov/public
moose-petsc               3.16.6                  build_3    https://conda.software.inl.gov/public
moose-tools               2022.11.07      py310ha2ead4d_0    https://conda.software.inl.gov/public
```
it could be that the update is failing, in which case you will need to rmeove the mamba environment and try again

For now, can you please try to make in moose/test and try the tests again

**Comment by joe61vette** [comment, marked as answer]
Thanks Guillaume.  That fixed the problem with not being able to run the Moose tests.  

I then went back and ran 'mamba update --all' and now have:

```bash
moose-libmesh             2022.11.07              build_0    https://conda.software.inl.gov/public
moose-libmesh-vtk         9.1.0               hf2274b2_14    https://conda.software.inl.gov/public
moose-mpich               4.0.2                   build_4    https://conda.software.inl.gov/public
moose-petsc               3.16.6                  build_3    https://conda.software.inl.gov/public
moose-tools               2022.11.07      py310hcffbc79_0    https://conda.software.inl.gov/public
```

So, should I rerun 'git clean -xfd' in my moose directory now that I have successfully updated the environment?

  Joe

**Reply by GiudGiud** [reply]
not every time really, only when you run into compilation issues.

**Comment by joe61vette** [comment]
So, now back to my original problem.  Just let me know if this should be a new issue.

In my app, I used 'git clean -xfd' and rebuilt.  I still get the dreaded "dyld[75408]: missing symbol called" when I try to run a test problem.

Is there any way to find what symbol?  

Thanks,
  Joe

**Reply by GiudGiud** [reply]
is this all you get?
Is this in opt mode or debug mode?

Since you git cleaned the rational explanation for that is now that there s a missing implementation in a class

**Comment by joe61vette** [comment]
It was in opt mode.  And sadly, that is all I get.  I ran in dbg mode earlier and didn't see anything useful either.

I'll go back and check every class for a missing function.  Likewise, I did an 'otool -L' and saw nothing meaningful.  Should I try it with dbg again?

Thanks for the help,
  Joe

**Reply by GiudGiud** [reply]
@lindsayad thoughts about this?

Can you please paste the otool -L output?

**Reply by GiudGiud** [reply]
@milljm

**Comment by joe61vette** [comment]
Here is the output:

```pre
/Users/joe/projects/ASP/asp-opt:
	/Users/joe/projects/ASP/test/lib/libasp_test-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/ASP/lib/libasp-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/modules/module_loader/lib/libmodule_loader_with_fp_ray_rdg_hc-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/modules/heat_conduction/lib/libheat_conduction-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/modules/rdg/lib/librdg-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/modules/ray_tracing/lib/libray_tracing-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/modules/fluid_properties/lib/libfluid_properties-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/framework/libmoose-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/framework/contrib/pcre/libpcre-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/joe/projects/moose/framework/contrib/hit/libhit-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/../libmesh/lib/libmesh_opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/../libmesh/lib/libtimpi_opt.11.dylib (compatibility version 12.0.0, current version 12.0.0)
	@rpath/libhdf5_cpp.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libvtkIOCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkCommonCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkCommonDataModel-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkFiltersCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkIOXML-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkImagingCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkIOImage-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkImagingMath-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkIOParallelXML-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkParallelMPI-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkParallelCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkCommonExecutionModel-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtksys-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.2.13)
	@rpath/libslepc.3.16.dylib (compatibility version 3.16.0, current version 3.16.2)
	@rpath/libpetsc.3.16.dylib (compatibility version 3.16.0, current version 3.16.6)
	@rpath/libHYPRE-2.23.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libstrumpack.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libsuperlu_dist.7.dylib (compatibility version 7.0.0, current version 7.1.1)
	@rpath/libhdf5_hl.200.dylib (compatibility version 201.0.0, current version 201.1.0)
	@rpath/libhdf5.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libparmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmpifort.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libpmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	@rpath/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libmpicxx.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
```

**Reply by milljm** [reply]
It looks like moose exists up one directory from your application. I see that `git clean -xfd` was run in previous posts. But I am not sure where it was performed. That said, when obtaining new versions of moose-libmesh (which means anything downstream of that; moose-petsc, moose-mpich), it is necessary to clean your build, rebuild, and then you can test. For you that means:

```bash
cd ~/projects/moose
git clean -xfd
cd ~/projects/ASP
git clean -xfd
make
```
Of course, be careful using `git clean -xfd`. This will delete any uncommitted, non-tracked files.

That said, did you perform a git clean in both the moose directory and your application directory? The links you provide all look good. But then again, the files won't change names, but the symbols in those file might (moose-libmesh's `@rpath/../libmesh/lib/libmesh_opt.0.dylib` library will have been updated if moose-libmesh was updated.)

**Comment by joe61vette** [comment]
I will redo my "cleaning" just to make sure it was done in the right order.

I tried doing a non_unity build just in case that might turn up something.  Ironically, the error it found was in the framework:

```pre
In file included from /Users/joe/projects/moose/framework/src/postprocessors/CumulativeValuePostprocessor.C:10:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/CumulativeValuePostprocessor.h:12:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/GeneralPostprocessor.h:13:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/Postprocessor.h:12:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/OutputInterface.h:13:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/MooseTypes.h:13:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/ADReal.h:13:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/DualRealOps.h:16:
In file included from /Users/joe/projects/moose/framework/build/header_symlinks/SparseOps.h:13:
In file included from /Users/joe/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualsemidynamicsparsenumberarray.h:31:
In file included from /Users/joe/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualsemidynamicsparsenumberarray_decl.h:27:
/Users/joe/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualnumber_decl.h:35:10: fatal error: 'metaphysicl/ct_types.h' file not found
#include "metaphysicl/ct_types.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~
```
Is that anything I should worry about?

Thanks.

**Reply by milljm** [reply]
Any order works, so if you performed a git clean in both, at any time after a `mamba update --all` should suffice. However:

This actually looks like what you have installed via Conda, is out-of-sync with what version of moose you have cloned (this version of moose might need an older version of moose-libmesh).

I suppose our instructions should say something like:

- If you update moose then
  - you should also update Conda (mamba update --all)
- If you update Conda then
  - you should also update moose (git fetch; git pull)

**Reply by joe61vette** [reply]
Did the cleaning in the order suggested, just to make sure.  Rebuilt and same problem with 'missing symbol called'.

**Reply by milljm** [reply]
I see you've done a fetch/rebase... You should have the latest available everything then... what is your origin?

```bash
cd ~/projects/moose
git remote -v
```

**Comment by joe61vette** [comment]
Here it is:

(moose) moose % git remote -v
origin	https://github.com/idaholab/moose.git (fetch)
origin	https://github.com/idaholab/moose.git (push)

**Reply by joe61vette** [reply]
Just rebuilt Moose & reran all tests.  No failures.

**Reply by milljm** [reply]
I am not seeing an issue in how/what you've done thus far:

```bash
cd ~/projects/moose
git rebase origin/master
mamba activate moose
mamba update --all
git clean -xfd
cd ../ASP
git clean -xfd
make
```

This appears to have been done throughout the course of this thread. But the error:
```pre
In file included from /Users/joe/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualsemidynamicsparsenumberarray_decl.h:27:
/Users/joe/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualnumber_decl.h:35:10: fatal error: 'metaphysicl/ct_types.h' file not found
#include "metaphysicl/ct_types.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~
```
does indicate something out of sync. Either the version of moose at ~/projects/moose, or what version of moose-libmesh is installed in your Conda environment (but you have the latest version, so... there is no disputing that).

You are also using the 'right' moose according to your `otool -L asp-opt` readout. I am a bit stumped at the moment...

**Reply by lindsayad** [reply]
@joe61vette, if possible, can you try to respond within a thread? It's a little more difficult to follow when the discussion is carried across Answers. What is the output of `echo $LIBMESH_DIR`?

**Reply by lindsayad** [reply]
I'm also curious what the undefined symbol was

**Reply by joe61vette** [reply]
Here you go:

ASP % echo $LIBMESH_DIR
/Users/joe/mambaforge3/envs/moose/libmesh

The major problem is that I don't know what symbol is missing.  Here is the error message:

(moose) Tests % asp 1d-gravity.i 
dyld[2210]: missing symbol called
zsh: abort      /Users/joe/projects/ASP/asp-opt -i 1d-gravity.i

Earlier I tried a dbg version but no info as to the missing symbol there either.

Tomorrow, I will systematically remove routines from my app until I find the one that causes the dyld problem.  I can't think of anything else.

**Reply by joe61vette** [reply]
Good Morning All:  I removed all of the header and source files from my app except those generated automatically when you run stork.sh.  Recompiled and the simple_diffusion.i test problem ran.  So, I'll begin adding back my routines to find the culprit.

Thanks again for the help,
  Joe

**Reply by joe61vette** [reply]
It appears that I was missing a small utility function.  Something that got missed when switching branches. Thanks everyone for the help.  At least I did need it to get my Moose environment straightened out.  I absolutely have to remember to use "mamba update --all" instead of "conda update --all" .

**Reply by lindsayad** [reply]
Sorry @joe61vette that sounds like a painful process. It's really obnoxious that your system didn't emit information about the missing symbol. My only suggestion in the future if you run into that same problem would be to try to get access to a Linux system because they list undefined symbols in detail (albeit mangled, but it can be demangled easily using web tools like demangler.com) at link time. Mac lazily checks for undefined symbols during run-time and apparently sometimes doesn't give detailed info about what's missing.

I'm very glad you were able to get it worked out though!
